libvast/vast/flatbuffer.hpp

```diff
@@ -0,0 +1,350 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2021 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#pragma once
+
+#include "vast/fwd.hpp"
+
+#include "vast/chunk.hpp"
+
+#include <flatbuffers/flatbuffers.h>
+#include <fmt/core.h>
+
+#include <compare>
+
+namespace vast {
+
+/// A wrapper class around a FlatBuffers table that allows for sharing the
+/// lifetime with the chunk containing the table.
+/// @tparam Table The generated FlatBuffers table type.
+/// @tparam IsRootTable Whether the table is a root table, i.e., starts at the
+/// beginning of the chunk.
+template <class Table, bool IsRootTable = true>
+class flatbuffer final {
+public:
+  // -- member types and constants --------------------------------------------
+
+  template <class ParentTable, bool ParentIsRootTable>
+  friend class flatbuffer;
+
+  friend struct ::fmt::formatter<flatbuffer>;
+
+  /// Indicates whether to verify the FlatBuffers table on construction.
+  enum class verify {
+    yes, ///< Perform extensive buffer verification.
+    no,  ///< Skip extensive buffer verification.
+  };
+
+#if VAST_ENABLE_ASSERTIONS
+  static constexpr auto verify_default = verify::yes;
+#else
+  static constexpr auto verify_default = verify::no;
+#endif // VAST_ENABLE_ASSERTIONS
+
+  // -- constructors, destructors, and assignment operators -------------------
+
+  /// Constructs a ref-counted FlatBuffers root table that shares the
+  /// lifetime with the chunk it's constructed from.
+  /// @pre *chunk* must hold a valid *Table*.
+  [[nodiscard]] static caf::expected<flatbuffer>
+  make(const chunk_ptr& chunk, enum verify verify = verify_default) noexcept
```

Comment:
For root tables, would it make sense to use `chunk_ptr&&` as argument? Or is there a use case where we have to share the chunk with something else?

Reply:
I'll take a look at the call sites. At least we should be taking by value to allow the user to move.